packages/tailwindcss/src/utilities.test.ts

```diff
@@ -15149,4 +15149,66 @@ describe('custom utilities', () => {
       `),
     ).toThrowError(/should be alphanumeric/)
   })
+
+  test('custom utilities work with `@apply`', () => {
+    expect(
+      compileCss(
+        css`
+          @utility foo {
+            @apply flex flex-col underline;
+          }
+
+          @utility bar {
+            @apply z-10;
+
+            .baz {
+              @apply z-20;
+            }
+          }
+
+          @tailwind utilities;
+        `,
+        ['foo', 'hover:foo', 'bar'],
+      ),
+    ).toMatchInlineSnapshot(`
+      ".bar {
+        z-index: 10;
+      }
+
+      .bar .baz {
+        z-index: 20;
+      }
+
+      .foo {
+        flex-direction: column;
+        text-decoration-line: underline;
+        display: flex;
+      }
+
+      .hover\\:foo:hover {
+        flex-direction: column;
+        text-decoration-line: underline;
+        display: flex;
+      }"
+    `)
+  })
+
+  test('custom utilities with `@apply` causing circular dependencies should error', () => {
```

Comment:
We should probably add a test verifying that apply of other `@utility`s work in an `@utility` — not just for the circular dep error case.